Bits refactor

glesica opened the pull request:
Refactor to use 128 bit unsigned integers instead of vectors of 64 bit integers. This will drastically improve memory performance for operations like enumerating the genotypes on a landscape, finding the frontiers between basins of attractions, or enumerating neutral networks.

georgelesica-wf committed:
Refactor genotypes and landscape to use bit strings.

glesica committed:
Updates to type aliases and added new tests.

glesica committed:
Refactor neighbor functions and reorganize code.

glesica committed:
Add additional tests.

glesica committed:
Add additional tests.

glesica committed:
Add additional neighbor tests.

glesica committed:
Clean up neighbor tests and fix bugs.

glesica committed:
Refactor walks for new genotypes.

ahalwright committed:
Revised  src/Population.jl  to be compatible with bits-refactoring.
Wrote  test/population.jl  to do basic unit tests.

ahalwright committed:
Modified test/population.jl to also test zeros(::Population, ls::Landscape, count::Int64).

ahalwright committed:
Adapted  src/enumeration.jl  to bits-refactor.  Did not do thorough testing or revise test file.

ahalwright committed:
Merge branch 'bits-refactor' of github.com:glesica/NKLandscapes.jl into bits-refactor

glesica committed:
Revise walks and add tests.

glesica committed:
Add population tests to runtests script.

glesica committed:
Rename enumeration and metapopulation files.

glesica committed:
Update metapopulation and population and tests.

glesica committed:
Fix equality and hash implementations.

glesica committed:
Remove comment.

glesica committed:
Update mutation to use new genotypes and populations.

glesica committed:
Finish new tests for mutation.

glesica committed:
Update and expand tests and fix bugs.

glesica committed:
Reorganize mutation code.

glesica committed:
Fix fast functional tests.

ahalwright committed:
Revised test/fastnets.jl to test the refactored neighbors, walks, and neutral nets.
Revised src/neighbors.jl to fix the "orequal=true" option of fitter_neighbors.
Revised src/enumeration.jl to remove unnecessary argument to "neutralnet" function.
Added a TODO concerning the documentation in src/landscape.jl.
TODO:  Debug  neutralnet  and  neutralnets  for the case where a = 3.

glesica committed:
Fix test runner script.

glesica committed:
Fix minor issues.

glesica committed:
Get Kauffman tests running again.

glesica committed:
Merge branch 'bits-refactor' of git://github.com/glesica/NKLandscapes.jl into bits-refactor

ahalwright committed:
Revised the documentation of function makelinks() in src/landscape.jl.
Added a check on the value of a in src/landscape/nk.jl, src/landscape/nkp.jl, src/landscape/nkq.jl (only a = 2 is implemented at this time).
Cleaned up and tested   test/fastnets.jl.
Added  fastnets" as an option in  runtests.sh.

ahalwright committed:
Merge branch 'bits-refactor' of github.com:glesica/NKLandscapes.jl into bits-refactor
Merge is necessary so that I can push additional changes.

ahalwright committed:
Did some cleanup of src/enumeration.jl.  Removed obsolete exports and changed documentation to agree with code.

ahalwright commented:
George,

Thank you very much for doing this refactor!!  I will look at it and at
the codecov results tomorrow.

I am in agreement with using 128 bit unsigned integers for NKLandscapes.
 This limits us to N=128 when the arity is 2, but I don't see that as a
problem.

For CGP I am dubious that using a single 128-bit unsigned integer will
be sufficient to encode all circuits that we want to use.  So maybe
vectors of 128-bit unsigned integers?

In basins-refactor, I combined the files  src/peak_paths.jl, 
test/peak_paths.jl, and example/peak_paths.jl  into the
example/peak_paths.jl  file.  Let's hold off on merging this branch for
a while since I am currently working on these files.  I will push my
commits to this branch to GitHub tomorrow.

Alden

On Tue, Mar 8, 2016, at 18:43, George Lesica wrote:

> Refactor to use 128 bit unsigned integers instead of vectors of 64 bit
> integers. This will drastically improve memory performance for operations
> like enumerating the genotypes on a landscape, finding the frontiers
> between basins of attractions, or enumerating neutral networks.
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/glesica/NKLandscapes.jl/pull/5
> 
> -- Commit Summary --
> - Refactor genotypes and landscape to use bit strings.
> - Updates to type aliases and added new tests.
> - Refactor neighbor functions and reorganize code.
> - Add additional tests.
> - Add additional tests.
> - Add additional neighbor tests.
> - Clean up neighbor tests and fix bugs.
> - Refactor walks for new genotypes.
> - Revised  src/Population.jl  to be compatible with bits-refactoring.
> - Modified test/population.jl to also test zeros(::Population,
>   ls::Landscape, count::Int64).
> - Adapted  src/enumeration.jl  to bits-refactor.  Did not do thorough
>   testing or revise test file.
> - Merge branch 'bits-refactor' of github.com:glesica/NKLandscapes.jl
>   into bits-refactor
> - Revise walks and add tests.
> - Add population tests to runtests script.
> - Rename enumeration and metapopulation files.
> - Update metapopulation and population and tests.
> - Fix equality and hash implementations.
> - Remove comment.
> - Update mutation to use new genotypes and populations.
> - Finish new tests for mutation.
> - Update and expand tests and fix bugs.
> - Reorganize mutation code.
> - Fix fast functional tests.
> - Revised test/fastnets.jl to test the refactored neighbors, walks, and
>   neutral nets.
> - Fix test runner script.
> - Fix minor issues.
> - Get Kauffman tests running again.
> - Merge branch 'bits-refactor' of
>   git://github.com/glesica/NKLandscapes.jl into bits-refactor
> - Revised the documentation of function makelinks() in
>   src/landscape.jl.
> - Merge branch 'bits-refactor' of github.com:glesica/NKLandscapes.jl
>   into bits-refactor
> - Did some cleanup of src/enumeration.jl.  Removed obsolete exports and
>   changed documentation to agree with code.
> 
> -- File Changes --
> 
> ```
> M runtests.sh (44)
> D src/Genotype.jl (49)
> D src/Landscape.jl (145)
> D src/Mutation.jl (88)
> M src/NKLandscapes.jl (21)
> D src/Neighbors.jl (177)
> D src/Population.jl (35)
> A src/aliases.jl (44)
> R src/enumeration.jl (93)
> A src/genotype.jl (76)
> A src/landscape.jl (59)
> A src/landscape/nk.jl (25)
> A src/landscape/nkp.jl (34)
> A src/landscape/nkq.jl (27)
> R src/metapopulation.jl (36)
> A src/mutation/bitstring.jl (53)
> A src/mutation/bitwise.jl (58)
> A src/neighbors.jl (126)
> A src/population.jl (38)
> M src/selection/moran.jl (28)
> M src/selection/proportional.jl (28)
> M src/selection/tournament.jl (28)
> R src/walks.jl (99)
> M test/fastfunc.jl (119)
> M test/fastnets.jl (42)
> A test/genotype.jl (55)
> M test/kauffman.jl (10)
> A test/landscape.jl (28)
> A test/metapopulation.jl (43)
> A test/mutation.jl (71)
> A test/neighbors.jl (110)
> A test/oldunit.jl (355)
> A test/population.jl (35)
> M test/runtests.jl (3)
> M test/unit.jl (359)
> A test/walks.jl (75)
> ```
> 
> -- Patch Links --
> 
> https://github.com/glesica/NKLandscapes.jl/pull/5.patch
> https://github.com/glesica/NKLandscapes.jl/pull/5.diff
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/glesica/NKLandscapes.jl/pull/5

ahalwright committed:
Revised the selection routines propsel!, tournsel!, moransel! to create new versions of each genotype copied.
Without this change, the resulting population would have multiple instances of the same genotype object
which led to bugs when the population was mutated with bwmutate!.
Also removed the comments about inefficiency in src/selection/tournsel.jl.  Testing showed that
the implementation of tournament selection is reasonably efficient, and that most of the computational
time of doing selection in the NK environment is in the computation of fitnesses.

ahalwright committed:
Merge branch 'bits-refactor' of github.com:glesica/NKLandscapes.jl into bits-refactor

Conflicts:
	src/selection/proportional.jl
	src/selection/tournament.jl

glesica committed:
Fix reference error in selection algorithms.

glesica committed:
Add tags to ignore file.

glesica committed:
Add elite selection.

glesica committed:
Merge branch 'bits-refactor' of git://github.com/glesica/NKLandscapes.jl into bits-refactor

glesica committed:
Adjust tournament fitness parameter usage.

glesica committed:
Add genotye mutation methods.

glesica committed:
Add initial migration rework.

glesica committed:
Refactor and test implementation.

glesica committed:
Finish migration and tests.

glesica commented:
OK, the refactor is finally finished, with the exception of fixing whatever weird breakage we have in Julia 0.5. I took a look at the change log and had figured out where the problem was, but now I've forgotten. I'll look into that today and then merge this when it is fixed, hopefully this afternoon. After that we can do a release to Julia's package repository, put a bow on it for awhile (other than bug fixes), and look at refactoring the CGP library to be compatible.

glesica committed:
Fix 0.5.0 bug.

glesica committed:
Consolidate unit tests and add additional tests.

glesica committed:
Add additional tests.

glesica committed:
Pull in nets tests and clean up.

ahalwright commented:
George,

Thank you very much for getting this done.  I expect it took a lot
longer that either of us expected.

I am going to going to work on stating goals relating to my ideas on
neutrality and NK landscapes.  Maybe I can merge these ideas with my
first idea on Hebbian learning.

I am not sure how to handle my basins-refactor branch which was a branch
of bits refactor.  

It includes:  
- basins.jl    functions to return a disjoint sets data structure
  representing the basins of attraction and to return a list of the
  basins.  The disjoint sets data structure can be used to efficiently
  find the basin of attraction of any genotype.  Includes a test file.
- peak_paths.jl     functions to find the least cost path between any
  two peaks (local fitness optima) where cost is a combination of steps
  and changes in fitness  Includes a test function.
- genetic_algorithm:   implements a basic genetic algorithm with options
  for selection, mutation, initialization.

These function are related to my ideas on neutrality and NK
landscapes---and a better statement of goals is needed to see where to
go with them.  

In terms of software management, I see two options:

1) reset the upstream head to be master rather than bits-refactor (I
don't know how to do this).  basins.jl and genetic_algorithm.jl  are
reasonable to include in NKLandscapes.jl.  I am not so sure about
peak_paths.jl

2) put them into a separate repository in a similar way to what you did
with soc_int_experiment.

Either way is OK with me.

Cheyenne and I need to follow up on my two ideas on cultural
evolution----can we get out a quick paper on these topics?

Alden

On Sat, Mar 26, 2016, at 11:35, George Lesica wrote:

> OK, the refactor is finally finished, with the exception of fixing
> whatever weird breakage we have in Julia 0.5. I took a look at the change
> log and had figured out where the problem was, but now I've forgotten.
> I'll look into that today and then merge this when it is fixed, hopefully
> this afternoon. After that we can do a release to Julia's package
> repository, put a bow on it for awhile (other than bug fixes), and look
> at refactoring the CGP library to be compatible.
> 
> ---
> 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/glesica/NKLandscapes.jl/pull/5#issuecomment-201903876
